internals/detect_intent.py

```diff
@@ -228,42 +238,57 @@ def load_detected_feature(self, feature_id: Optional[int],
 
     return FeatureEntry.get_by_id(fe_ids[0]), None
 
-  def set_intent_thread_url(self, feature: FeatureEntry,
+  def find_matching_stage(self, feature: FeatureEntry,
       approval_field: approval_defs.ApprovalFieldDef,
-      thread_url: str | None, subject: str | None) -> None:
-    """If the feature has no previous thread URL for this intent, set it."""
+      thread_url: str | None) -> Stage | None:
+    """Find the stage in which the thread is associated with."""
     if not thread_url:
       logging.info('Given false thread_url %r', thread_url)
-      return
+      return None
 
     stage_type = core_enums.STAGE_TYPES_BY_GATE_TYPE_MAPPING[
         approval_field.field_id][feature.feature_type]
     if stage_type is None:
       logging.info('stage_type not found for %r %r',
                    approval_field.field_id, feature.feature_type)
-      return
-    # TODO(danielrsmith): A new way to approach this detection is needed
-    # now that multiple stages of the same type can exist for a feature.
-    # This will currently only detect an intent if there is only 1 stage
-    # of the specific stage type associated with the feature.
-    matching_stages: list[Stage] = Stage.query(
+      return None
+
+    # Get all stages of the detected stage type that belong to the feature.
+    stages_of_type_in_feature: list[Stage] = Stage.query(
         Stage.feature_id == feature.key.integer_id(),
         Stage.stage_type == stage_type).fetch()
-    if len(matching_stages) == 0 or len(matching_stages) > 1:
-      logging.info('Ambiguous stages: %r', matching_stages)
-      return
-    if matching_stages[0].intent_thread_url:
+    if len(stages_of_type_in_feature) == 0:
+      logging.info('No matching stage found for feature: '
+                   f'{feature.key.integer_id()}, stage_type: {stage_type}')
+      return None
+
+    # Check if this intent URL already belongs to a stage.
+    matching_stage = next((s for s in stages_of_type_in_feature
```

Comment:
I don't think that there will ever be a match because the `thread_url` is actually contains the SMTP Message-ID value of a specific message within the thread like `/d/msgid/blink-dev/MESSAGE-ID`.  The Message-ID of the first message in a thread will not == the Message-ID of later messages in the same thread.  We use the URL of the first message in the thread to represent the whole thread, but we never store a URL for the whole thread.

I think this code would often work anyway, simply because when a feature owner creates a new origin trial, and then posts anything to a new or old "Intent to experiment" thread about that feature, they are probably talking about the new origin trial or more-or-less ready to review it.  Worst case, the review shows up on the reviewers' dashboard too soon.

To make this work precisely, I think that we would need to change the URL that is embedded in the intent email to include `?gate=GATE_ID`.  That could be a helpful change for all intents anyway.

Reply:
I see. I agree that it's probably best for us to use the gate ID query string to make sure we're obtaining the correct gate. I can update this logic and add the query string in a subsequent PR. 🙂

Reply:
Adding a TODO here, but will follow up ASAP